alns/weights/WeightScheme.py

```diff
@@ -48,17 +52,26 @@ def select_operators(self, rnd_state: RandomState) -> Tuple[int, int]:
         rnd_state
             Random state object, to be used for number generation.
 
+        op_coupling
+            Matrix that indicates coupling between destroy and repair
+            operators. Entry (i, j) is 1 if destroy operator i can be used in
+            conjunction with repair operator j and 0 otherwise.
+
         Returns
         -------
         A tuple of (d_idx, r_idx), which are indices into the destroy and repair
         operator lists, respectively.
         """
 
-        def select(op_weights):
+        def select(op_weights, op_indices):
             probs = op_weights / np.sum(op_weights)
-            return rnd_state.choice(range(len(op_weights)), p=probs)
+            return rnd_state.choice(op_indices, p=probs)
+
+        d_idx = select(self._d_weights, self._d_indices)
+        selection = np.flatnonzero(op_coupling[d_idx])
+        r_idx = select(self._r_weights[selection], self._r_indices[selection])
 
-        return select(self._d_weights), select(self._r_weights)
+        return d_idx, r_idx
```

Comment:
Right. I had something similar before but it looked very ugly because I had a very long name for what is now called `selection`. 

Which brings me back to this point. Do you think its better to rename `selection` to `r_idcs` or `r_selection` to improve readability?

Reply:
Suggestion: call `selection` `coupled_r_ops`. I think that makes it _much_ clearer.

Reply:
It fits - see the new commit [45234d9](https://github.com/N-Wouda/ALNS/pull/70/commits/45234d9196e89bfd41e1599bd85130fed6368efb).